File: src/main/java/com/lessmarkup/engine/data/dialects/MySqlDialect.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.lessmarkup.engine.data.dialects;

import java.util.OptionalInt;

public class MySqlDialect implements DatabaseLanguageDialect {

    @Override
    public String getTypeDeclaration(DatabaseDataType type, OptionalInt sizeLimit, boolean required, String characterSet) {
        String nullable = required ? " NOT NULL" : " NULL";
        switch (type){
            case INT:
                return "INT" + nullable;
            case LONG:
                return "BIGINT" + nullable;
            case DATE_TIME:
                return "DATETIME" + nullable;
            case STRING:
                if (characterSet == null) {
                    characterSet = "utf8";
                }
                if (!sizeLimit.isPresent()) {
                    return "TEXT CHARACTER SET " + characterSet + nullable;
                }
                return String.format("VARCHAR(%s) CHARACTER SET " + characterSet, sizeLimit.getAsInt()) + nullable;
            case BOOLEAN:
                return "BIT" + nullable;
            case FLOAT:
                return "REAL" + nullable;
            case DOUBLE:
                return "DOUBLE" + nullable;
            case BINARY:
                return "BLOB" + nullable;
            case IDENTITY:
                return "BIGINT" + nullable + " AUTO_INCREMENT";
            default:
                throw new IllegalArgumentException();
        }
    }

    @Override
    public String decorateName(String name) {
        return "`" + name + "`";
    }

    @Override
    public DatabaseDataType getDataType(String dataType) {
        switch (dataType.toUpperCase()) {
            case "INT":
                return DatabaseDataType.INT;
            case "BIGINT":
                return DatabaseDataType.LONG;
            case "DATETIME":
                return DatabaseDataType.DATE_TIME;
            case "TEXT":
            case "VARCHAR":
                return DatabaseDataType.STRING;
            case "BIT":
                return DatabaseDataType.BOOLEAN;
            case "REAL":
                return DatabaseDataType.FLOAT;
            case "DOUBLE":
                return DatabaseDataType.DOUBLE;
            case "BLOB":
                return DatabaseDataType.BINARY;
            default:
                return null;
        }
    }

    @Override
    public String paging(int from, int count) {
        return String.format("LIMIT %d,%d", from, count);
    }
}

File: src/main/scala/com/lessmarkup/engine/data/dialects/MySqlDialect.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.data.dialects

class MySqlDialect extends DatabaseLanguageDialect {
  def getTypeDeclaration(dataType: DatabaseDataType, sizeLimit: Option[Int], required: Boolean, characterSet: Option[String]): String = {
    val nullable: String = if (required) " NOT NULL" else " NULL"
    dataType match {
      case DatabaseDataType.INT => "INT" + nullable
      case DatabaseDataType.LONG => "BIGINT" + nullable
      case DatabaseDataType.DATE_TIME => "DATETIME" + nullable
      case DatabaseDataType.STRING =>
        if (sizeLimit.isEmpty)
          "TEXT CHARACTER SET " + characterSet.getOrElse("utf8") + nullable
        else
          s"VARCHAR(${sizeLimit.get}) CHARACTER SET ${characterSet.getOrElse("utf8")} $nullable"
      case DatabaseDataType.BOOLEAN => "BIT" + nullable
      case DatabaseDataType.FLOAT => "REAL" + nullable
      case DatabaseDataType.DOUBLE => "DOUBLE" + nullable
      case DatabaseDataType.BINARY => "BLOB" + nullable
      case DatabaseDataType.IDENTITY => "BIGINT AUTO_INCREMENT"
      case _ => throw new IllegalArgumentException
    }
  }

  def decorateName(name: String): String = {
    "`" + name + "`"
  }

  def getDataType(dataType: String): DatabaseDataType = {
    dataType.toUpperCase match {
      case "INT" => DatabaseDataType.INT
      case "BIGINT" => DatabaseDataType.LONG
      case "DATETIME" => DatabaseDataType.DATE_TIME
      case "TEXT" => DatabaseDataType.STRING
      case "VARCHAR" => DatabaseDataType.STRING
      case "BIT" => DatabaseDataType.BOOLEAN
      case "REAL" => DatabaseDataType.FLOAT
      case "DOUBLE" => DatabaseDataType.DOUBLE
      case "BLOB" => DatabaseDataType.BINARY
      case _ => null
    }
  }

  def paging(from: Int, count: Int): String = {
    s"LIMIT $from,$count"
  }

  override def schemaTablesName: String = "INFORMATION_SCHEMA.TABLES"

  override def definePrimaryKey: Boolean = true
}
